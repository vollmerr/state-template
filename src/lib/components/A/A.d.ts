import * as React from 'react';

export interface AProps {
  /** Path for internal links, mutually exlusive with `href` */
  to?: string;

  /** Path for external links (open in new tab), mutually exlusive with `to` */
  href?: string;

  /** Content to render, mutually exlusive with `children` */
  text?: string;

  /** Content to render */
  children?: React.ReactNode;
}

declare class A extends React.Component<AProps, {}> {}

export default A
