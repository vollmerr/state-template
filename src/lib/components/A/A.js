import React from 'react';
import T from 'prop-types';
import { Link } from 'react-router-dom';

import Icon from '../Icon';

/**
 * Anchor tag that handles internal and external links.
 * Opens external links in a new window.
 */
const A = (props) => {
  const {
    children,
    href,
    iconProps,
    text,
    to,
    ...rest
  } = props;

  let aProps;
  let Tag = 'a';
  if (href) {
    // is external link
    aProps = {
      ...rest,
      href,
      target: '_blank',
      rel: 'noopener noreferrer',
    };
  } else {
    // is internal link
    aProps = {
      ...rest,
      to,
    };
    Tag = Link;
  }

  let content;
  if (text) {
    content = text;
  } else if (children) {
    content = children;
  } else if (iconProps) {
    const { srOnly, ...icon } = iconProps;
    content = (
      <>
        <Icon {...icon} />
        {srOnly && <span className={'sr-only'}>{srOnly}</span>}
      </>
    );
  }

  return (
    <Tag data-test={'link'} {...aProps}>
      {content}
    </Tag>
  );
};

A.propTypes = {
  /** Render content */
  children: T.node,

  /** Path for external links (open in new tab) */
  href: T.string,

  /** Props for icon to render as content */
  iconProps: T.shape(Icon.propTypes),

  /** Render content using shorthand */
  text: T.string,

  /** Path for internal links */
  to: T.string,
};

A.defaultProps = {
  children: null,
  href: '',
  iconProps: null,
  text: '',
  to: '',
};

export default A;
