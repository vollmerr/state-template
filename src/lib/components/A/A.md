### internal

```jsx
<p>This is an <A to={'/testPath'}>internal anchor</A></p>
```

### external
External links open in a new tab.

```jsx
<p>This is an <A href={'http://www.google.com'}>external anchor</A></p>
```

### shorthand
You can use the shorthand `text` prop instead of passing `children`.

```jsx
<p>This is an <A to={'/testPath'} text={'shorthand anchor'} /></p>
```

### icon link
You can render the links as an [Icon](#icon) using the `iconProps`.

```jsx
<A iconProps={{ name: 'home' }} />
```
