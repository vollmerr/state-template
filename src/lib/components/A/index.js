import React from 'react';
import T from 'prop-types';
import { Link } from 'react-router-dom';

/**
 * Anchor tag that handles internal and external links
 * opens external links in a new window
 */
const A = (props) => {
  const {
    to,
    href,
    text,
    children,
    ...rest
  } = props;

  let aProps;
  let Tag = 'a';
  if (href) {
    // is external link
    aProps = {
      ...rest,
      href,
      target: '_blank',
    };
  } else {
    // is internal link
    aProps = {
      ...rest,
      to,
    };
    Tag = Link;
  }

  return <Tag {...aProps}>{text}</Tag>;
};

A.propTypes = {
  /** Path for internal links */
  to: T.string,
  /** Path for external links (open in new tab) */
  href: T.string,
  /** Content to display as a link */
  text: T.string.isRequired,
};

A.defaultProps = {
  to: '',
  href: '',
};

export default A;
