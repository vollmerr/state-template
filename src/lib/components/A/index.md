### internal

```jsx
  <p>This is an <A to={'/testPath'} text={'internal anchor'} /></p>
```

### external
External links open in a new tab.

```jsx
  <p>This is an <A href={'http://www.google.com'} text={'external anchor'} /></p>
```
