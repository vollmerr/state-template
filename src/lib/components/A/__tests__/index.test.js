import React from 'react';
import { shallow } from 'enzyme';
import { Link } from 'react-router-dom';

import { A as fromLib } from '../../..';
import A from '..';

const props = {
  href: 'http://www.google.com',
  text: 'link text',
};

let wrapper;
describe('A', () => {
  beforeEach(() => {
    wrapper = shallow(<A {...props} />);
  });

  it('should be exported as part of the library', () => {
    expect(fromLib).toEqual(A);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render children if no `text`', () => {
    wrapper.setProps({ text: '', children: 'children...' });
    expect(wrapper.text()).toEqual('children...');
  });

  describe('external links', () => {
    it('should use an `a` tag', () => {
      const tag = wrapper.type();
      expect(tag).toBe('a');
    });

    it('should open in a new tab', () => {
      const target = wrapper.prop('target');
      expect(target).toBe('_blank');
    });
  });

  describe('internal links', () => {
    it('should use a `Link` tag', () => {
      wrapper.setProps({ to: '/testPath', href: undefined });
      const tag = wrapper.type();
      expect(tag).toBe(Link);
    });
  });

  it('should pass additional props', () => {
    wrapper.setProps({ id: 'testId', type: 'submit', variant: 'primary' });
    expect(wrapper.prop('id')).toBe('testId');
    expect(wrapper.prop('type')).toBe('submit');
    expect(wrapper.prop('variant')).toBe('primary');
  });
});
