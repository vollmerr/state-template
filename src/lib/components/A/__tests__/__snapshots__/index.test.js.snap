// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A should render correctly 1`] = `
<a
  href="http://www.google.com"
  target="_blank"
>
  link text
</a>
`;
